### PR TITLE
WIP insp3 update/rewrite

### LINKED ISSUE #84: Update for 3.0
The current build only supports 2.x, but 3.0 was released recently and should be dockerized. I'm not very familiar with the intricacies of Inspircd and Docker, but I'll try to get at least a start on it and see if I can find some of the gotchas.

### PR DESCRIPTION
**Pull request by r734**
## Checklist

- [x] Implementation
- [ ] Tests
- [ ] Docs

Fixes #84 

This is an update for insp3. Tests/Docs/CI are not done yet, but the image builds with a simple configuration and generates a certificate on startup if needed.

I'm new to inspircd and open source contribution in general, so please let me know if I'm doing something incorrectly.

### CODE REVIEW

**Comment by r734** [comment]
Okay, I'm looking more into the configuration. Here's the workflow for what I'm planning (and as I think about it more, it may not actually be different from the current setup):

- During image build, clone and compile the original repo
- Remove the original repo's config at /inspircd/run/conf/*
- Copy the docker repo's config files to /inspircd/conf/
  - (At this point the docker repo's config files are in the image but not set to be used)

Then a user pulls and runs the image...
- The user can volume mount a custom config with `-v /path/to/conf:/inspircd/conf`; if they do, then the config file(s) in the container from the build step are effectively overwritten
- If /inspircd/run/conf is empty (i.e. this is a fresh image), the entrypoint.sh script will run `cp -r /inspircd/conf /inspircd/run/conf`
  - If the user didn't volume mount a config, this effectively makes the docker repo's config files be the ones that run
    - From there, the .sh files that are (will be) included in inspircd.conf will run and handle config generation from env variables and secrets as needed
    - If a user doesn't provide any environment vars, sane defaults will be used

Is that accurate?

**Comment by SISheogorath** [comment]
That sounds very good, yes and pretty much reassembles the current flow.

There is just one difference and I think we can figure that one out: We don't include the source code into the final version. Instead we build inspircd with a different prefix and then remove the source repository later on.

https://github.com/inspircd/inspircd-docker/blob/a61807e36879c1ce45fc321de44ff2da4e7928fb/Dockerfile#L32-L33

Which will change the config path from `/inspircd/run/conf` to `/inspircd/conf`.

Maybe let me give you some insight into the current docker image's config workflow (which is quite similar to your idea):

There is `/conf` and `/inspircd/conf`. `/conf` is a docker volume while `/inspircd/conf` is not.

On startup the entrypoint script goes like this:

https://github.com/inspircd/inspircd-docker/blob/a61807e36879c1ce45fc321de44ff2da4e7928fb/entrypoint.sh#L4-L21

So when `/inspircd/conf` exists and we can write to it, we'll just fix the ownership and copy over the configs from the image (including the `config.sh`). That config is then ready for the admin to modify because bind mounts always start empty which is rather annoying.

If we can't write to `/inspircd/conf` it'll just show a warning. As it's up to the admin to do things there.

If we `/inspircd/conf` doesn't exist, we link `/inspircd/conf` to `/conf` which is a volume and therefore will not only keep config changes done by CLI commands like `docker exec` but also keep the generated certificates unless someone deletes the volume.

Why not making `/inspircd/conf` the volume? Because you can't copy configs form there into a bind-mount in same position, and because you can't write to this path anymore in images that are based on this one.

So in order to allow admins to add their own modifed configs to an own image, they can just extent it by using: 

```Dockerfile
FROM inspircd/inspircd:2.0

COPY ./myconfigs /inspircd/conf

```

Which wouldn't work when `/inspircd/conf` is a volume.

---

Sorry for the wall of text, I simply put quite some thought into this config structure and even when it might seems over-engineered, I wanted to share this knowledge with you.

Important: We don't have to setup it exactly that way again, so feel free to say "okay, that's too weird".

File: Dockerfile [hunk 1]
```diff
@@ -1,75 +1,50 @@
...
-VOLUME ["/conf"]
+RUN git clone https://github.com/inspircd/inspircd.git inspircd
 
-WORKDIR /inspircd/
+WORKDIR /inspircd
+RUN git checkout $(git describe --abbrev=0 --tags $VERSION)
+
+## TODO add module support here
+
+RUN ./configure $CONFIGUREARGS --disable-interactive --uid 10000 --gid 10000
```
Comment (SaberUK) [inline Dockerfile]: You don't need to pass --disable-interactive if you're passing other configure flags.

**Comment by SISheogorath** [comment]
Oh I just realized that due to the builder pattern we no longer have the problem of having the source code around :see_no_evil: all good, please go ahead then!

**Comment by r734** [comment]
Good news...I'm able to use this image as a base for another image with a custom config file. For instance, say I build this image:

`docker build -t inspircd3 .`

Then create another folder with two files:

inspircd.conf
```
<bind address="" port="6668" type="clients">
```

Dockerfile
```
FROM inspircd3
COPY inspircd.conf /inspircd/conf/
EXPOSE 6668
```

Then build this new image with `docker build -t myinspircd3 .`, and finally run with `docker run -d -p 6668:6668 myinspircd3`, it works as expected: clients can connect unencrypted on port 6668.

Furthermore, I can stop the container and start it back up again, and it Just Works without clobbering anything.

So at this point I'll start working on reimporting and updating the shell scripts to generate configuration from environment variables and secrets.

Then I need to learn about and deal with modules (may need some help to know exactly what to do/change).

**Comment by SISheogorath** [comment]
Amazing work! I really look forward to merge it :)

File: conf/setup.sh [hunk 2]
```diff
@@ -0,0 +1,5 @@
+#!/bin/sh
+
+cat <<EOF
+<define name="hostname" value="${HOSTNAME:-irc}">
```
Comment (SISheogorath) [inline conf/setup.sh]: At least in insp20 this has to be an FQDN. So maybe `irc.example.com`.

Comment (SaberUK) [reply conf/setup.sh]: It has to be one in v3 too.

**Comment by r734** [comment]
See #86 for a replacement PR - I overcomplicated this, thinking that I would end up rewriting all of the config from the ground up...But it turns out the config is (or at least seems?) mostly compatible with insp3.

**Pull request closed without merging**